File: pkg/bus/dlq.go
package bus

import (
	"context"
	"fmt"

	"github.com/ModulationAI/openagentio/pkg/codec"
	"github.com/ModulationAI/openagentio/pkg/event"
	"github.com/ModulationAI/openagentio/pkg/middleware"
	"github.com/ModulationAI/openagentio/pkg/transport"
)

// DLQSink returns a middleware.DLQSink that publishes a cloned envelope onto
// the dead-letter subject `{prefix}.dlq.{event_type}` using the supplied codec
// and transport. This is the canonical NATS Core–based DLQ implementation; it
// does not depend on JetStream.
//
// The sink stamps two metadata keys on the clone before publishing:
//   - acp.dlq.original_event_type — the original envelope.EventType
//   - acp.dlq.last_error            — the string representation of lastErr
func DLQSink(prefix string, codec codec.Codec, tr transport.Transport) middleware.DLQSink {
	return func(ctx context.Context, e *event.Envelope, lastErr error) error {
		cp := e.Clone()
		if cp.Metadata == nil {
			cp.Metadata = make(map[string]any)
		}
		cp.Metadata["acp.dlq.original_event_type"] = cp.EventType
		if lastErr != nil {
			cp.Metadata["acp.dlq.last_error"] = lastErr.Error()
		}
		data, err := codec.EncodeEnvelope(cp)
		if err != nil {
			return fmt.Errorf("dlq: encode envelope: %w", err)
		}
		subject := prefix + ".dlq." + cp.EventType
		return tr.Publish(ctx, &transport.RawMessage{
			Subject: subject,
			Data:    data,
		})
	}
}

File: sdk/python/src/openagentio/bus/dlq.py
"""DLQSink factory. Mirrors pkg/bus/dlq.go.

Returns a :data:`DLQSink` that publishes a cloned envelope onto the
dead-letter subject ``{prefix}.dlq.{event_type}``. The sink stamps two
metadata keys on the clone:

- ``acp.dlq.original_event_type`` — the original EventType
- ``acp.dlq.last_error`` — the string representation of lastErr
"""
from __future__ import annotations

from openagentio.bus.subjects import DEFAULT_SUBJECT_PREFIX
from openagentio.codec.json_codec import Codec
from openagentio.event.envelope import Envelope
from openagentio.middleware.deadletter import DLQSink
from openagentio.transport.base import RawMessage, Transport


def dlq_sink(
    prefix: str = DEFAULT_SUBJECT_PREFIX,
    codec: Codec | None = None,
    transport: Transport | None = None,
) -> DLQSink:
    """Create a DLQSink that publishes to ``{prefix}.dlq.{event_type}``.

    The caller must supply *transport* (and optionally *codec*).
    """
    if transport is None:
        raise ValueError("dlq_sink: transport is required")
    c = codec or __import__("openagentio.codec.json_codec", fromlist=["JSONCodec"]).JSONCodec()

    async def sink(env: Envelope, last_err: Exception) -> None:
        cp = env.clone()
        if cp.metadata is None:
            cp.metadata = {}
        cp.metadata["acp.dlq.original_event_type"] = cp.event_type
        if last_err is not None:
            cp.metadata["acp.dlq.last_error"] = str(last_err)
        data = c.encode_envelope(cp)
        subject = f"{prefix}.dlq.{cp.event_type}"
        await transport.publish(RawMessage(subject=subject, data=data))

    return sink